docs(linter): add recipe for flat config converter (#18614)
nrwl/nx: docs/shared/recipes/tips-n-tricks/migrating-to-flat-eslint.md

# Switching to ESLint's flat config format

Version 8 of ESLint introduced a new configuration format called [Flat Config](https://eslint.org/docs/latest/use/configure/configuration-files-new). The next major version will use this config format by default. The purpose of this format is to:

- push towards a single configuration format (in contrast to the existing `JSON`, `Yaml` and `JS`-based configs)
- enforce explicit native loading (instead of the implicit imports in `JSON` and `Yaml`)
- use a flat cascading of rules (instead of a mix of rules and overrides)

> See below direct comparison between `JSON`, `JS` and `Flat` config:

Reviewer: ```suggestion
See below a direct comparison between `JSON`, `JS` and `Flat` config:
```

{% tabs %}

> {% tab label="eslint.config.js" %}

Author: I wouldn't change this to Flat. We can add it in parentheses.

> ```js

Reviewer: ```suggestion
```js {% fileName="eslint.config.js" %}
```

const js = require('@eslint/js');
const baseConfig = require('./eslint.base.config.js');
const globals = require('globals');
const jsoncParser = require('jsonc-eslint-parser');
const tsParser = require('@typescript-eslint/parser');

module.exports = [
  js.configs.recommended,
  // this will spread the export blocks from the base config
  ...baseConfig,
  { plugins: { '@nx': nxPlugin } },
  {
    languageOptions: {
      parser: tsParser,
      globals: {
        ...globals.node,
      },
    },
    rules: {
      '@typescript-eslint/explicit-module-boundary-types': ['error'],
    },
  },
  // there are no overrides, all the config blocks are "flat"
  {
    files: ['*.json'],
    languageOptions: {
      parser: jsoncParser,
    },
    rules: {},
  },
  {
    files: ['*.ts', '*.tsx', '*.js', '*.jsx'],
    rules: {
      '@nx/enforce-module-boundaries': [
        'error',
        {
          enforceBuildableLibDependency: true,
          allow: [],
          depConstraints: [
            {
              sourceTag: '*',
              onlyDependOnLibsWithTags: ['*'],
            },
          ],
        },
      ],
    },
  },
];
```

{% /tab %}

> {% tab label=".eslintrc.json" %}

Reviewer: ```suggestion
{% tab label="JSON" %}
```

> ```json

Reviewer: ```suggestion
```json {% fileName=".eslintrc.json" %}
```

{
  "root": true,
  "parser": "@typescript-eslint/parser",
  "env": {
    "node": true
  },
  "extends": ["eslint:recommended", "./.eslintrc.base.json"],
  "plugins": ["@nx"],
  "rules": {
    "@typescript-eslint/explicit-module-boundary-types": "error"
  },
  "overrides": [
    {
      "files": ["*.json"],
      "parser": "jsonc-eslint-parser",
      "rules": {}
    },
    {
      "files": ["*.ts", "*.tsx", "*.js", "*.jsx"],
      "rules": {
        "@nx/enforce-module-boundaries": [
          "error",
          {
            "enforceBuildableLibDependency": true,
            "allow": [],
            "depConstraints": [
              {
                "sourceTag": "*",
                "onlyDependOnLibsWithTags": ["*"]
              }
            ]
          }
        ]
      }
    }
  ]
}
```

{% /tab %}

> {% tab label=".eslintrc.js" %}

Reviewer: ```suggestion
{% tab label="JS" %}
```

> ```js

Reviewer: ```suggestion
```js {% fileName=".eslintrc.js" %}
```

> # Switching to the ESLint's flat config format

Reviewer 2: ```suggestion
# Switching to ESLint's Flat Config format
```

> Version 8 of ESLint introduced a new configuration format called [Flat Config](https://eslint.org/docs/latest/use/configure/configuration-files-new). The purpose of this format is to:

Reviewer 2: Maybe we should also say it will be the default in v9?

> - propose a single configuration format (in contrast to the existing `JSON`, `Yaml` and `JS`-based configs)

Reviewer: Instead of `propose`, I think you mean `create` or `provide`.  Its not a proposal

Author: I actually meant `propose` but I understand the confusion. I will change it.

- enforce explicit native loading (instead of the implicit imports in `JSON` and `Yaml`)

> - use flat cascading of rules (instead of a mix of rules and overrides)

Reviewer: ```suggestion
- use a flat cascading of rules (instead of a mix of rules and overrides)
```

> For additional details, head over to their [official blog post](https://eslint.org/blog/2022/08/new-config-system-part-2/).

Reviewer: ```suggestion
For additional details, head over to ESLint's [official blog post](https://eslint.org/blog/2022/08/new-config-system-part-2/).
```

I wasn't sure who you meant by "their"

Since version 16.7.0, Nx supports usage of flat config in the [@nx/linter:eslint](/packages/linter/executors/eslint) executor and `@nx/*` generators and provides an automated config conversion from `.eslintrc.json` config files.

## Converting workspace from .eslintrc.json to flat config

> To convert workspace ESLint configurations from the default `.eslintrc.json` to new flat config you need to run:

Reviewer: ```suggestion
To convert workspace ESLint configurations from the default `.eslintrc.json` to the new flat config you need to run:
```

```shell
 nx g @nx/linter:convert-to-flat-config
```

The generator will go through all the projects and convert their configurations to the new format. It will also convert the base `.eslintrc.json` and `.eslintignore`.

## Correctness and best practices

> The purpose of this generator is to create a flat config that works the same way, as the original `JSON` config did. Depending on the complexity of your original config, it may be using `FlatCompat` for providing a compatibility wrapper around parts of the original config. You can improve those by following the [official migration guide](https://eslint.org/docs/latest/use/configure/migration-guide).

Reviewer 2: ```suggestion
The purpose of this generator is to create a flat config that works the same way, as the original `JSON` config did. Depending on the complexity of your original config, it may be using the `FlatCompat` utility to provide a compatibility wrapper around parts of the original config. You can improve those by following the [official migration guide](https://eslint.org/docs/latest/use/configure/migration-guide).
```

Reviewer: Remove the comma here: `same way, as the original`
